### PR TITLE
FC-ize InteractionMasks

### PR DESCRIPTION
**Pull request by amanmahajan7**
Notable changes
- I have not migrated `selectedRange` functionality in the PR as it requires some cleanup and bug fixing
- Added `scrollToCell` function. 
- `Canvas` no longer needs `col*Idx` props

I am still working on fixing remaining tests but the PR can be reviewed

### CODE REVIEW

File: src/masks/InteractionMasks.tsx [hunk 1]
```diff
@@ -174,64 +153,58 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
-        this.handlePaste();
-      }
-    }
+  function closeEditor(): void {
+    setEnableEditorAfterRender(false);
+    setIsEditorEnabled(false);
+    setFirstEditorKeyPress(null);
   }
 
-  onFocus = (): void => {
```
Comment (amanmahajan7) [inline src/masks/InteractionMasks.tsx]: Not sure if we need `onFocus`

File: src/Cell.tsx [hunk 2]
```diff
@@ -27,10 +27,8 @@ function Cell<R>({
   onRowDoubleClick,
   enableCellRangeSelection
 }: CellProps<R>) {
-  const position: Position = { idx, rowIdx };
-
   function selectCell(openEditor?: boolean) {
-    eventBus.dispatch(EventTypes.SELECT_CELL, position, openEditor);
+    eventBus.dispatch(EventTypes.SELECT_CELL, { idx, rowIdx }, openEditor);
```
Comment (amanmahajan7) [inline src/Cell.tsx]: I changed this so selecting the same cell again would trigger [focus](https://github.com/adazzle/react-data-grid/blob/af1676eec78ba7433885869289a6769466ce0c38/src/masks/InteractionMasks.tsx#L107).

File: src/masks/InteractionMasks.tsx [hunk 3]
```diff
@@ -172,66 +120,96 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
+
+  // Focus on the selection mask after the editor is closed
+  useEffect(() => {
+    if (isEditorEnabled) return;
+    selectionMaskRef.current?.focus();
+  }, [isEditorEnabled]);
+
+  useEffect(() => {
+    if (!enableEditorAfterRender || !isCellEditable) return;
+    setEnableEditorAfterRender(false);
```
Comment (nstepien) [inline src/masks/InteractionMasks.tsx]: Shouldn't we call this whether the cell is editable or not?

File: src/masks/InteractionMasks.tsx [hunk 4]
```diff
@@ -172,66 +120,96 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
+  useEffect(() => {
+    return eventBus.subscribe(EventTypes.SELECT_CELL, selectCell);
+  });
+
+  useEffect(() => {
+    if (!draggedPosition) return;
+    const handleDragEnter = (overRowIdx: number) => {
+      setDraggedPosition({ ...draggedPosition, overRowIdx });
+    };
+    eventBus.subscribe(EventTypes.DRAG_ENTER, handleDragEnter);
```
Comment (nstepien) [inline src/masks/InteractionMasks.tsx]:
```suggestion
    return eventBus.subscribe(EventTypes.DRAG_ENTER, handleDragEnter);
```

File: src/masks/InteractionMasks.tsx [hunk 5]
```diff
@@ -599,110 +349,84 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
-    const { columns, rowHeight } = this.props;
-    return (
-      <>
-        <SelectionRangeMask
-          {...getSelectedRangeDimensions({ selectedRange: this.state.selectedRange, columns, rowHeight })}
-        />
-        <SelectionMask
-          {...this.getSelectedDimensions(this.state.selectedPosition)}
-          ref={this.selectionMask}
+  const rowData = rowGetter(selectedPosition.rowIdx);
```
Comment (nstepien) [inline src/masks/InteractionMasks.tsx]: Can we always expect `rowGetter` to handle `-1` index cases?

Comment (amanmahajan7) [reply src/masks/InteractionMasks.tsx]: We cannot. I have moved it so it is only called when the editor is enabled

File: src/masks/InteractionMasks.tsx (hunk 3 above)
Comment (amanmahajan7) [reply src/masks/InteractionMasks.tsx]: Fixed

File: src/Canvas.tsx [hunk 6]
```diff
@@ -113,33 +105,34 @@ function Canvas<R, K extends keyof R>({
...
-      const { clientWidth, scrollLeft } = current;
-      const isCellAtLeftBoundary = left < scrollLeft + width + frozenColumnsWidth;
+      const isCellAtLeftBoundary = left < scrollLeft + width + getFrozenColumnsWidth();
       const isCellAtRightBoundary = left + width > clientWidth + scrollLeft;
       if (isCellAtLeftBoundary || isCellAtRightBoundary) {
         const newScrollLeft = getColumnScrollPosition(columns, idx, scrollLeft, clientWidth);
-        if (newScrollLeft !== 0) {
-          current.scrollLeft = scrollLeft + newScrollLeft;
-        }
+        current.scrollLeft = scrollLeft + newScrollLeft;
```
Comment (amanmahajan7) [inline src/Canvas.tsx]: I ended up using `scrollLeft` and `scrollTop` props. They should have the same value I believe

File: src/masks/InteractionMasks.tsx [hunk 7]
```diff
@@ -599,110 +351,82 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
-
-  onCommitCancel = (): void => {
-    this.closeEditor();
-  };
+    closeEditor();
+  }
 
-  getRowTop(rowIdx: number): number {
-    return rowIdx * this.props.rowHeight;
+  function onCommitCancel(): void {
```
Comment (nstepien) [inline src/masks/InteractionMasks.tsx]: I think we can remove this function and use `closeEditor` directly.

File: src/masks/InteractionMasks.tsx [hunk 8]
```diff
@@ -172,66 +120,98 @@ export default class InteractionMasks<R, K extends keyof R> extends React.Compon
...
+
+    if (!isCellEditable) return;
+    setEditorPosition(getEditorPosition());
+  }, [enableEditorAfterRender, getEditorPosition, isCellEditable]);
+
+  useEffect(() => {
+    if (enableCellAutoFocus && document.activeElement === document.body) {
+      setSelectedPosition({ idx: 0, rowIdx: 0 });
+    }
+  }, [enableCellAutoFocus]);
```
[the remark this answers is not in the document]
Comment (amanmahajan7) [reply src/masks/InteractionMasks.tsx]: Looks good 👍

**Pull request merged**